### PR TITLE
useSettings: extract selector

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

This PR splits out the selector from `useSettings`, so it can be combined with other `useSelect` selectors in a single subscription.

## Why?

There's a use case for inner blocks, reducing the blockEditor store subscriptions from 2 to 1, but there's also used cases in some blocks.

First run -10% type.

ALSO, the useSettings hook cannot be called conditionally, but the selector can. So you can do:

```js
useSelect( ( select ) => condition ? select( blockEditorStore ).getBlockSettings( clientId, 'something' ) : null, ... )
```

I've seen a few places where that could be useful to avoid even more subscriptions.

## How?

I created a new `getBlockSettings` private selector. Not sure if the naming is correct.

## Testing Instructions

E2e tests should pass.

### CODE REVIEW

File: packages/block-editor/src/components/use-settings/index.js [hunk 1]
```diff
@@ -143,120 +25,15 @@ export function hasMergedOrigins( value ) {
...
-			} );
-		},
-		[ blockName, clientId, paths ]
+		( select ) =>
+			unlock( select( blockEditorStore ) ).getInheritedSettings(
+				clientId,
+				...paths
+			),
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		[ clientId, ...paths ]
```
Comment (ellatrix) [inline packages/block-editor/src/components/use-settings/index.js:36]: Whether we spread the dependencies here or in `useMemo`, it should be the same.

File: packages/block-editor/src/components/global-styles/effects-panel.js [hunk 2]
```diff
@@ -26,7 +26,7 @@ import { shadow as shadowIcon, Icon, check } from '@wordpress/icons';
 /**
  * Internal dependencies
  */
-import { mergeOrigins } from '../use-settings';
+import { mergeOrigins } from '../../store/private-selectors/get-inherited-settings';
```
Comment (ellatrix) [inline packages/block-editor/src/components/global-styles/effects-panel.js]: Maybe this should be moved to a utils file, but I left it for now, there's only two uses.

File: packages/block-editor/src/store/index.js [hunk 3]
```diff
@@ -10,11 +10,17 @@ import reducer from './reducer';
...
+import { getInheritedSettings } from './get-inherited-settings';
 import * as resolvers from './resolvers';
 import * as actions from './actions';
 import { STORE_NAME } from './constants';
 import { unlock } from '../lock-unlock';
 
+privateSelectors = {
+	...privateSelectors,
+	getInheritedSettings,
+};
```
Comment (jsnajdr) [inline packages/block-editor/src/store/index.js]: Let's reexport `getInheritedSettings` in the `private-selectors` module, that's more elegant.

File: packages/block-editor/src/store/get-inherited-settings.js [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+ * @param {Object} value Object to check
+ * @return {boolean} Whether the object has values in any of the three origins
+ */
+export function hasMergedOrigins( value ) {
+	return [ 'default', 'theme', 'custom' ].some(
+		( key ) => value?.[ key ]?.length
+	);
+}
+
+export function getInheritedSettings( state, clientId, ...paths ) {
```
Comment (jsnajdr) [inline packages/block-editor/src/store/get-inherited-settings.js]: I don't like the name very much, `getBlockSettings` is probably better. But it doesn't matter that much for a private API.

**Review by jsnajdr — APPROVED** [review]
Another nice idea :+1:

File: packages/block-editor/src/store/index.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-editor/src/store/index.js]: For some reason that causes the unit test to fail, I haven't figured out why

File: packages/block-editor/src/store/get-inherited-settings.js (hunk 4 above)
Comment (ellatrix) [reply packages/block-editor/src/store/get-inherited-settings.js]: Changed it

**Pull request merged**